datastore/zmq/enabled.go

```diff
@@ -0,0 +1,188 @@
+//go:build has_zmq
+
+package zmq
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/pebbe/zmq4"
+	"github.com/sirupsen/logrus"
+	"github.com/teslamotors/fleet-telemetry/metrics"
+	"github.com/teslamotors/fleet-telemetry/metrics/adapter"
+	"github.com/teslamotors/fleet-telemetry/telemetry"
+)
+
+// Metrics stores metrics reported from this package
+type Metrics struct {
+	errorCount   adapter.Counter
+	publishCount adapter.Counter
+	byteTotal    adapter.Counter
+}
+
+var (
+	metricsRegistry Metrics
+	metricsOnce     sync.Once
+)
+
+// ZMQProducer implements the telemetry.Producer interface by publishing to a
+// bound zmq socket.
+type ZMQProducer struct {
+	topicPrefix string
+	topicSuffix string
+	topicSep    string
+	ctx         context.Context
+	sock        *zmq4.Socket
+	logger      *logrus.Logger
+}
+
+// Publish the record to the socket.
+func (p *ZMQProducer) Produce(rec *telemetry.Record) {
+	if p.ctx.Err() != nil {
+		return
+	}
+
+	topicParts := make([]string, 0, 4)
+	if len(p.topicPrefix) > 0 {
+		topicParts = append(topicParts, p.topicPrefix)
+	}
+	topicParts = append(topicParts, rec.Vin)
+	topicParts = append(topicParts, rec.TxType)
+	if len(p.topicSuffix) > 0 {
+		topicParts = append(topicParts, p.topicSuffix)
+	}
+	joinChar := DefaultTopicSep
+	if len(p.topicSep) > 0 {
+		joinChar = p.topicSep
+	}
+	topic := strings.Join(topicParts, joinChar)
```

Comment:
for better readability/testability, maybe extract topic generation logic to a new method and then add a unit test for it? Or try reusing https://github.com/teslamotors/fleet-telemetry/blob/394063b718fc1306a7e57945f250ed5bac447996/telemetry/producer.go#L24